Feat/dashboard

f-allian opened the pull request:
- Added a lot of code to `dashboards.py`
- Polished up the `SurveyView` to render metrics for the dashboard
- Added a filter functionality to allow users to filter by demographic.



## Example Dashboard:

![image](https://github.com/user-attachments/assets/a5e21e31-a817-423e-8ab5-dca01b43928b)


## Example with Filter:


![image](https://github.com/user-attachments/assets/3c4e969d-3b69-4ddb-ba6c-d6ec1751388b)

twinkarma commented on survey/dashboards.py:
```diff
@@ -0,0 +1,1708 @@
+        ], className='g-4', style={'marginBottom': '20px'}),
+
+        # Section A
+        dbc.Row([
```
Maybe use a loop for make a separate function to generate each section? There's repetition of code for all the sections.

twinkarma commented on survey/dashboards.py:
```diff
@@ -0,0 +1,1708 @@
+            ], width=12)
+        ], style={'marginTop': '20px', 'marginBottom': '20px'}),
+
+        dbc.Row([
```
A function can be used to generate the configuration of each of the demographic section

twinkarma commented on survey/dashboards.py:
```diff
@@ -0,0 +1,1708 @@
+    return fig
+
+
+@dash_app.callback(
```
There must be a better way of listing these callbacks but let's discuss it after the user review.

twinkarma reviewed (changes requested):
## Review for tomorrow
The interface is looking good. I've made some suggestions in the code itself but since it doesn't affect what's being rendered, we can address it after the user review tomorrow.

As for what should be fixed for tomorrow:

- Maturity score should be between 0-4

![image](https://github.com/user-attachments/assets/47651d27-5325-46dd-81dc-25976b8e7526)

- The "Early progress" bar is very faint, it's hard to see especially in the first section when it's all squashed together:

![image](https://github.com/user-attachments/assets/bee31bf1-e38f-4d18-b639-4a0cf2e9dbe0)

- Ethnicity pie chart is cut off by the legend (this is when it's showing all the results without any filters selected)
![image](https://github.com/user-attachments/assets/e9f6699f-6aa6-4da3-9332-3ab1cbce4ee1)


## Not a priority for tomorrow 
But I'm making a note here so we don't forget it:

- We should add a spinner when filters change and we're waiting for the data to load.
- For filtering/comparison it's probably going to be more useful if we can somehow show the table of the demograhic categories at the same time e.g.
![image](https://github.com/user-attachments/assets/dfce8094-6239-47be-a1a7-ea31209e151d)

twinkarma commented:
What's dash's support like for generating a pdf. If there isn't do you want me to look at at this for tomorrow? I was going to do an excel export function as well.

twinkarma commented:
@f-allian I've already done the csv export in #131 

Will look into pdf generation now.

f-allian commented:
@twinkarma Thanks for your comments, I've fixed them above. Should we merge this into dev now?

f-allian commented:
@twinkarma I've addressed your comments above and have improved the overall structure of the dashboard functionality. Main changes are:

- I've removed the massive `dashboard.py` script and changed it into an app
- The dashboard now gets the data from the SurveyResponse model instead of the config files
- There is now an "Age Group" filter option.
- Added a loading spinner while the dashboard updates.

Read for another review, thanks 👍🏼

twinkarma commented on survey/dashboards/config.py:
```diff
@@ -0,0 +1,124 @@
+
+# Demographic fields configuration
+DEMOGRAPHIC_FIELDS = [
+    {
```
What happens when someone changes the order of questions or add a demographic question out of order?

twinkarma commented on survey/dashboards/config.py:
```diff
@@ -0,0 +1,124 @@
+
+# Survey sections configuration
+SECTIONS = [
+    {
```
Might be better to just loop through the sections?

twinkarma commented on survey/dashboards/config.py:
```diff
@@ -0,0 +1,124 @@
+from django.conf import settings
+
+
+SORT_CONFIG_PATH = os.path.join(
```
Config should probably be based on individual survey.survey_config rather than the JSON file?

twinkarma commented on survey/dashboards/config.py:
```diff
@@ -0,0 +1,124 @@
+    settings.BASE_DIR, "data", "survey_config", "sort_only_config.json"
+)
+DEMOGRAPHIC_CONFIG_PATH = os.path.join(
+    settings.BASE_DIR, "data", "survey_config", "demography_only_config.json"
```
Should be based on either survey.suvey_config or survey.demography_config instead? As user may change the configuration.

twinkarma reviewed (changes requested):
We'll need to make sure that the dashboard is configured to that particular survey instead of the global json file. If the manage changes the demographic questions or if there's a case where SORT itself changes (e.g. additional sections added, names changed for specific professions, etc.) then the dashboard will break/display the wrong information when trying to view older surveys.

twinkarma commented on survey/dashboards/dashboard.py:
```diff
@@ -0,0 +1,21 @@
+from .callbacks import register_callbacks
+from survey.models import SurveyResponse
+
+survey_data = list(SurveyResponse.objects.all().values())
```
Just noticed this but this call essentially gets all response data from our entire database, all responses from different surveys will be mixed together!

f-allian commented:
@twinkarma I've committed my new changes that answer the dashboard's configuration/permission issues above. I've also included the functionality to toggle a drop-down for users to view qualitative comments, as mentioned in #147:


### Closed:
![image](https://github.com/user-attachments/assets/f69d01fd-b6ae-4ad1-9c6d-d9d356fb414a)

### Open:
![image](https://github.com/user-attachments/assets/f783424b-13e9-4583-9b7b-db59bf1eb05f)

Joe-Heffer-Shef commented on survey/dashboards/callbacks.py:
```diff
@@ -0,0 +1,308 @@
+        [Input(f'{field["id"]}-filter', "value") for field in demographic_fields],
+    )
+    def update_stored_data(*filter_values):
+        all_responses = getattr(dash_app, 'survey_responses', [])
```
I think this should be
```python
all_responses = dash_app.survey_responses
```
because there shouldn't be a situation where the Dash app is initialised without that attribute. If there is, it should raise an exception.

To me, it seems more intuitive if we have a class structure that lets us do
```python
dash_app.survey.responses
```

Joe-Heffer-Shef commented on survey/dashboards/callbacks.py:
```diff
@@ -0,0 +1,221 @@
+    )
+    def update_stored_data(*filter_values):
+
+        all_responses = getattr(dash_app, 'survey_responses', [])
```
I think that the Dash app has it's own HTTP API endpoints that serve up responses based on the registered callbacks. For example:
```
[04/Apr/2025 11:39:44] "POST /django_plotly_dash/app/SurveyDashboard_3/_dash-update-component HTTP/1.1" 200 25742
[04/Apr/2025 11:39:44] "POST /django_plotly_dash/app/SurveyDashboard_3/_dash-update-component HTTP/1.1" 200 860
[04/Apr/2025 11:39:44] "POST /django_plotly_dash/app/SurveyDashboard_3/_dash-update-component HTTP/1.1" 200 304
```
The `dash_app.update_stored_data()` callback takes all the survey responses and applies the filters, then returns that filtered sequence of responses and the aggregated results.

All the survey responses are stored in the `DjangoDash`  instance as an attribute. That data stays in the backend and the filtered survey data is sent via that endpoint.

Joe-Heffer-Shef commented on survey/dashboards/callbacks.py:
```diff
@@ -0,0 +1,308 @@
+            if field["id"] == "age":
+                options.append([
+                    {"label": opt, "value": opt}
+                    for opt in ["Under 25", "25-34", "35-44", "45-54", "55-64", "65+"]
```
Can we define these values in one place to avoid repetition?
![image](https://github.com/user-attachments/assets/c693653e-b2cd-432c-9848-ea08d2f088e0)

Joe-Heffer-Shef commented on survey/dashboards/callbacks.py:
```diff
@@ -0,0 +1,308 @@
+                f"{avg_score:.1f}",
+            )
+
+        except Exception as e:
+            print(f"Error in update_metrics: {str(e)}")
+            return "0", "0%", "0.0"
```
How come we're capturing all exceptions? Exceptions should be specific for situations that can be handled, otherwise let them bubble up.

Joe-Heffer-Shef commented on survey/dashboards/callbacks.py:
```diff
@@ -0,0 +1,308 @@
+    for section in sections:
+        # Section breakdown graph
+        @dash_app.callback(
+            Output(f'section-breakdown-{section["letter"].lower()}', "figure"),
```
Don't repeat the same code
```python
for section in sections:
  letter = section["letter"].lower()
...
```

Joe-Heffer-Shef commented on survey/dashboards/config.py:
```diff
@@ -0,0 +1,107 @@
+
+        print(f"Extracted {len(sections)} sections from survey_config")
+        return sections
+    except Exception as e:
```
Remove general exception

Joe-Heffer-Shef commented on survey/dashboards/config.py:
```diff
@@ -0,0 +1,107 @@
+
+from .utils import get_age_group
+
+def get_sections_from_survey(survey):
```
Should this be a method of the `Survey` class?

Joe-Heffer-Shef commented on survey/dashboards/config.py:
```diff
@@ -0,0 +1,107 @@
+                    })
+
+        return demographic_fields
+    except Exception as e:
```
Remove general exception

Joe-Heffer-Shef commented on survey/dashboards/config.py:
```diff
@@ -0,0 +1,107 @@
+        return []
+
+
+def get_demographic_fields_from_survey(survey):
```
Should this be a method of the `Survey` class? e.g. `Survey.demographic_fields`

Joe-Heffer-Shef commented on survey/dashboards/constants.py:
```diff
@@ -0,0 +1,125 @@
+}
+
+# Colour palette and maturity levels
+COLOUR_PALETTE = [
```
Some of these constants should probably `tuple`s because they want to be immutable.

Joe-Heffer-Shef commented on survey/dashboards/constants.py:
This is all good, but be aware that there's a few ways to make little data structures that can be useful to avoid bugs by putting more meaning into the code such as dataclass, namedtuple, enums, etc. Another way to do this is using classes like so:
```python
class Theme:
    PRIMARY = "#1192e8"
    PRIMARY_BACKGROUND = "#e5f6ff"
...
```

Joe-Heffer-Shef commented on survey/dashboards/dashboard.py:
```diff
@@ -0,0 +1,53 @@
+from .config import get_sections_from_survey, get_demographic_fields_from_survey
+
+
+def get_survey_dashboard(survey_id):
```
```python
def get_survey_dashboard(survey_id: int) -> DjangoDash:
...
```

Joe-Heffer-Shef commented on survey/dashboards/dashboard.py:
```diff
@@ -0,0 +1,53 @@
+
+        return dash_app
+
+    except Exception as e:
```
Remove general exception.

Joe-Heffer-Shef commented on requirements.txt:
```diff
@@ -27,7 +27,11 @@ wcwidth==0.2.13
 gunicorn==23.*
 strenum==0.4.15
+django-plotly-dash===2.4.3
+numpy==2.1.3
```
Why do we need these (big) libraries?

Joe-Heffer-Shef commented on requirements.txt:
```diff
@@ -27,7 +27,11 @@ wcwidth==0.2.13
 pytest-django==4.9.0
 django-qr-code==4.1.0
 factory-boy==3.3.3
+scipy==1.15.1
```
Why do we need scipy?

Joe-Heffer-Shef commented on survey/dashboards/utils.py:
On line 238, we're using NumPy
```python
section_scores.append(np.mean(valid_scores))
```
Do we need this big scientific library?
Why not use `statistics.mean` that's built in to Python? https://docs.python.org/3/library/statistics.html

Joe-Heffer-Shef commented on survey/templates/survey/report.html:
Where is this file used?

Joe-Heffer-Shef commented on survey/views.py:
```diff
@@ -20,37 +22,95 @@
 
 
 class SurveyView(LoginRequiredMixin, View):
+    login_url = '/login/'
```
Attributes should come after the docstring.

Joe-Heffer-Shef commented on survey/views.py:
```diff
@@ -20,37 +22,95 @@
 
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.survey_service = SurveyService()
```
This doesn't feel right. The survey service is initialised in the `survey.services` module already. I think should be in the global scope like this instead, it's at the service "layer" of the app:

```python
from .services import survey_service
```
There are references to this `survey_service` global variable in this view, so I don't think this code will work as it is?

Joe-Heffer-Shef commented on survey/views.py:
```diff
@@ -20,37 +22,95 @@
-    def post(self, request: HttpRequest, pk: int):
-        return self.render_survey_page(request, pk, is_post=True)
+    def post(self, request, pk):
+        print("POST method called")
```
Remove print statements

Joe-Heffer-Shef commented on survey/views.py:
```diff
@@ -20,37 +22,95 @@
-        context["can_edit"] = {
-            survey.id: survey_service.can_edit(request.user, survey)
+    def get_survey_metrics(self, responses):
+        section_averages = {}
+        response_list = []
```
Minor quibble, but in general it's more readable to use the full data type names
```python
section_averages = dict()
response_list = list()
```
to avoid confusion e.g. is `{}` a `set` or a `dict`?

Joe-Heffer-Shef commented on survey/views.py:
```diff
@@ -20,37 +22,95 @@
+            }
+            response_list.append(response_data)
+
+            for section_idx, section_values in enumerate(answers):
```
Should these calculations be implemented as a method on `Survey`?

Joe-Heffer-Shef commented on survey/views.py:
```diff
@@ -20,37 +22,95 @@
+    responses = survey.survey_response.all()
+    context["has_responses"] = responses.exists()
+    if context["has_responses"]:
+        get_survey_dashboard(survey.id)  # Only get the survey for a given id
```
What does this line do?
